ste/sender-blockBlob.go

```diff
@@ -168,22 +169,28 @@ func (s *blockBlobSenderBase) Cleanup() {
 	jptm := s.jptm
 
 	// Cleanup
-	if jptm.TransferStatus() <= 0 { // TODO: <=0 or <0?
-		// If the transfer status value < 0, then transfer failed with some failure
+	if jptm.IsDeadInflight() {
 		// there is a possibility that some uncommitted blocks will be there
 		// Delete the uncommitted blobs
-		// TODO: should we really do this deletion?  What if we are in an overwrite-existing-blob
-		//    situation. Deletion has very different semantics then, compared to not deleting.
 		deletionContext, cancelFn := context.WithTimeout(context.Background(), 30*time.Second)
 		defer cancelFn()
-		_, _ = s.destBlockBlobURL.Delete(deletionContext, azblob.DeleteSnapshotsOptionNone, azblob.BlobAccessConditions{})
-		// TODO: question, is it OK to remoe this logging of failures (since there's no adverse effect of failure)
-		//  if stErr, ok := err.(azblob.StorageError); ok && stErr.Response().StatusCode != http.StatusNotFound {
-		// If the delete failed with Status Not Found, then it means there were no uncommitted blocks.
-		// Other errors report that uncommitted blocks are there
-		// bbu.jptm.LogError(bbu.blobURL.String(), "Deleting uncommitted blocks", err)
-		//  }
-
+		if jptm.WasCanceled() {
+			// If we cancelled, and the only blocks that exist are uncommitted, then clean them up.
+			// This prevents customer paying for their storage for a week until they get garbage collected, and it
+			// also prevents any issues with "too many uncommitted blocks" if user tries to upload the blob again in future.
+			// But if there are committed blocks, leave them there (since they still safely represent the state before our job even started)
+			blockList, err := s.destBlockBlobURL.GetBlockList(deletionContext, azblob.BlockListAll, azblob.LeaseAccessConditions{})
+			hasUncommittedOnly := err == nil && len(blockList.CommittedBlocks) == 0 && len(blockList.UncommittedBlocks) > 0
+			if hasUncommittedOnly {
+				jptm.LogAtLevelForCurrentTransfer(pipeline.LogDebug, "Deleting uncommitted destination blob due to cancellation")
+				// Delete can delete uncommitted blobs.
+				_, _ = s.destBlockBlobURL.Delete(deletionContext, azblob.DeleteSnapshotsOptionNone, azblob.BlobAccessConditions{})
+			}
+		} else {
+			// TODO: review (one last time) should we really do this?  Or should we just give better error messages on "too many uncommitted blocks" errors
```

Comment:
>"too many uncommitted blocks" [](start = 110, length = 29)

I think if we want to log this we should check the error that causes the failure (or check the uncommitted block list length)

Reply:
Yes, that makes sense to check the error code if we're going to log it something.  There are pros and cons of deleting here always.  Pros are that its consistent, both with itself (always does some thing regardless of uncommitted block count) and with Files and ADLS Gen2); and that it automatically handles the too many uncommitted blocks issue, without customer intervention.  The con is that its (usually) unnecessary to delete.  

I think I'm still leaning towards the following: It's easier and clearer to explain that "AzCopy always deletes the destination object if there's failure" (without having to get into the subtleties of it doing something different for Blob compared to everything else.  I think that's the same logic that brought us to the deletion approach in the first place back before GA.

Also, that's what it has been doing since GA. (While cancellation behavior is fixed by this PR, behavior on failure of a single transfer has always been to delete, even for blob).  The last thing we'd want to do is change it in this release, to something different from what it's historically been.... and then change our minds and change it back.   It makes more sense to leave it as it always has been, and only change it if/when we are certain.

So think I'm still leaning, slightly, towards we should indeed delete here.

Let's see what @zezha-msft says...